### ISSUE TITLE
glibc mbrtowc() function fails when locale is C or POSIX; this breaks the read builtin command

### ISSUE THREAD

**Issue by TimWolla** [fish-shell/fish-shell#2802]
Ctrl + Shift + T opens a new tab in gnome-terminal. The current working directory in this new tab usually is the same as the one in the previously active tab. But it is not when the current working directory contains a German umlaut.

### Reproduction Steps:

```
# Create directory with umlaut and change into it
cd /tmp/
mkdir Tür
cd Tür/
# Now: Press Ctrl+Shift+T
```

### Expected behavior:

`pwd` is `/tmp/Tür`

### Observed behavior:

`pwd` is `/home/<myuser>`

### Additional information:

---
Fish version: `2.2.0-675-g5e09411`

Operating system: Ubuntu Wily with fish's nightly PPA

Terminal or terminal emulator: `GNOME Terminal 3.16.2`

**Comment by krader1961** [comment]
The commit ID in that fish version isn't in the official source repo. So I can't easily tell what version you're running. The problem you're reporting should have been fixed by commit 1767681f9a61ac0789f5dafa76126e3446924b6d on 2015-12-12 and subsequently improved upon by commit d8a497434f8c22e698c049332b301b4ef4815c21 on 2016-02-17. Can you confirm whether your version includes those changes?

**Comment by TimWolla** [comment]
> The commit ID in that fish version isn't in the official source repo.

It is. It's this commit: https://github.com/fish-shell/fish-shell/commit/5e09411340b347caf2f81dc65fc9d3a1fefac261 which is the `HEAD` of `master`. The `g` cannot be part of the hash (for obvious reasons) and stands for “git”.

**Comment by krader1961** [comment]
> The g cannot be part of the hash...

Doh! Of course. Need more caffeine.

Any chance you have access of a system running Mac OS X? If so can you confirm this works correctly using OS X Terminal.app? It does for me using your example path. I'm guessing there's something about gnome terminal's handling of this feature that is different from Terminal.app. Perhaps a slightly different encoding is required. I'll try to reproduce on Ubuntu when I get home in a few hours.

**Comment by TimWolla** [comment]
> Any chance you have access of a system running Mac OS X?

Unfortunately I don't. It works correctly when entering `gnome-terminal` instead of opening a new tab using Ctrl+Shift+T, though.

**Comment by faho** [comment]
I can't actually get this feature working or find documentation for how exactly this needs to be encoded, but vte's `__vte_osc7` bash function definitely encodes ümläüts differently - "/tmp/ü" yields:

```bash
0000000 1b 5d 37 3b 66 69 6c 65 3a 2f 2f 66 6f 72 64 70  >.]7;file://fordp<
0000020 72 65 66 65 63 74 2f 74 6d 70 2f 25 43 33 25 42  >refect/tmp/%C3%B<
0000040 43 07                                            >C.<
0000042
```

in bash and

```fish
0000000 1b 5d 37 3b 66 69 6c 65 3a 2f 2f 66 6f 72 64 70  >.]7;file://fordp<
0000020 72 65 66 65 63 74 2f 74 6d 70 2f 07              >refect/tmp/.<
0000034
```

in fish.

The called functions are defined as

```bash
__vte_urlencode() (
  # This is important to make sure string manipulation is handled
  # byte-by-byte.
  LC_ALL=C
  str="$1"
  while [ -n "$str" ]; do
    safe="${str%%[!a-zA-Z0-9/:_\.\-\!\'\(\)~]*}"
    printf "%s" "$safe"
    str="${str#"$safe"}"
    if [ -n "$str" ]; then
      printf "%%%02X" "'$str"
      str="${str#?}"
    fi
  done
)

__vte_osc7 () {
  printf "\033]7;file://%s%s\007" "${HOSTNAME:-}" "$(__vte_urlencode "${PWD}")"
}
```

**Comment by krader1961** [comment]
@faho, your fish debug output shows your $PWD was */fordprefect/tmp/* rather than the directory with an umlaut in its name. What I see in my reproduction is fish is encoding the umlaut as `%CC%88` rather than `%C3%BC` as done by bash. The former is wrong, the latter is correct. Both the original fix by @jakwings and the subsequent rewrite by @kballard produce the same encoding.

Bingo! The difference is in how the non-ASCII char is UTF-8 encoded in **$PWD**. This is bash:

```
$ cd /private/tmp/Tür/
$ echo $PWD | od -tx1z
0000000 2f 70 72 69 76 61 74 65 2f 74 6d 70 2f 54 c3 bc  >/private/tmp/Tü<
0000020 72 0a                                            >r.<
0000022
```

This is fish:

```
$  cd /tmp/Tür/
$  echo $PWD | od -tx1z
0000000 2f 70 72 69 76 61 74 65 2f 74 6d 70 2f 54 75 cc  >/private/tmp/Tu�<
0000020 88 72 0a                                         >.r.<
0000023
```

Notice the difference in how the u+umlaut is represented. This reflects a difference in how bash and fish handle combining characters. Mac OS X Terminal.app treats both representations equivalently whereas Gnome terminal does not.

I am not a Unicode expert so I'm going to disengage and leave this for someone who is a Unicode expert to figure out what, if anything, should be changed with respect to how fish handles Unicode.

**Comment by krader1961** [comment]
I've changed the label from "bug" to "needs-more-info" because I can't tell from the analysis done to this point whether fish is correctly or incorrectly UTF-8 encoding strings that contain combining diacritical marks. I strongly suspect what fish is doing is strictly speaking "correct". However, if in practice that causes problems and there is an alternative that works in more situations then fish should probably be changed. Which would mean this isn't really a bug but rather an enhancement request.

**Comment by faho** [comment]
>@faho, your fish debug output shows your $PWD was /fordprefect/tmp/ rather than the directory with an umlaut in its name.

It actually wasn't - I just tested it again, and _with_ $PWD = "/tmp/ü" it still prints that output. It seems that on my system it's completely missing the umlauts. When I create "/tmp/äpfel", it's printing "/tmp/pfel".

**Comment by krader1961** [comment]
@faho, then it seems you've found another problem but probably one that is external to fish. It looks like your debug output was collected using the `echo $PWD | od -tx1z` I used. Yes or no? What is the output of `locale`?

**Comment by kballard** [comment]
CC88 is the UTF-8 encoding for U+0308 COMBINING DIARESIS, which is the combining mark that combines with "u" to produce "ü". C3BC is the UTF-8 encoding for U+00FC LATIN SMALL LETTER U WITH DIARESIS. They both represent the same grapheme, but the former is the decomposed version and the latter is the precomposed version.

@faho Can you provide the output of `ls -1 | xxd`? This should show what the actual encoding of the pathname is without the shell's involvement.

**Comment by faho** [comment]
>It looks like your debug output was collected using the echo $PWD | od -tx1z I used. Yes or no? What is the output of locale?

`pwd | __fish_urlencode | od -t x1z`. LC_ALL is unset, all other locale variables (LANG, LC_NAME etc) are "de_DE.UTF-8".

>@faho Can you provide the output of ls -1 | xxd? This should show what the actual encoding of the pathname is without the shell's involvement.

It's C3BC.

**Comment by krader1961** [comment]
@kballard: On OS X I created a directory then did

```
$ touch Tür
$ ls -1 | xxd
00000000: 5475 cc88 720a                           Tu..r.
```

Which suggests that what fish is doing is not obviously wrong, since another program produces the same encoding as fish, although it might be suboptimal from a compatibility viewpoint. I also tested this with @faho's locale and got the same result.

**Comment by kballard** [comment]
@krader1961 OS X's filesystem (HFS+) is unicode-aware and actually converts pathnames into decomposed form automatically when storing them. So it's impossible to create a directory that has U+00FC in it, you always get the version with U+0308.

**Comment by krader1961** [comment]
@kballard: Yep, I just recognized something like that must be happening because I logged into my Ubuntu server, performed the same experiment, and saw *xxd* report the C3BC hex sequence.

**Comment by kballard** [comment]
When `cd`ing into a directory in Bash, it seems to preserve the filename you gave it as the PWD. On OS X, because the filesystem does the decomposition for you, you can type either the precomposed or the decomposed path and you'll get the same directory. In Bash if I `cd` into the precomposed version, `pwd` reports the precomposed pathname, and if I `cd` into the decomposed version, `pwd` reports the decomposed pathname. But Fish always reports the decomposed pathname. My vague recollection is that Fish is actually retrieving the path from the current directory (e.g. from a file handle) instead of merely preserving whatever it was given.

You can see this much more easily with symlinks. If you `cd` into a symlink, e.g. `/tmp`, Bash reports the path you gave it for the `pwd`, but Fish reports the "real" path that the symlink pointed to (e.g. `/private/tmp`).

**Comment by krader1961** [comment]
I don't think your symlink example is relevant, @kballard, but the rest of your analysis looks correct to me. So the question remains: Is fish behaving according to accepted best practices with respect to Unicode characters that have combining diacritical marks and therefore the problem lies with Gnome terminal? Note that OS X Terminal.app does the right thing with both shells in control.

**Comment by jakwings** [comment]
Way to reproduce on Ubuntu (15.10):

```bash
sudo add-apt-repository ppa:fish-shell/nightly-master
sudo apt-get update
sudo apt-get install fish

fish --version  #=> fish, version 2.2.0-675-g5e09411
```

In fish:

```fish
echo Tür | __fish_urlencode | xxd
````

outputs

```
00000000: 5472                                     Tr
```

I modified `__fish_urlencode` and added `echo $chars | xxd` but got this:

```
00000000: 5472 7220 0a0a                           Tr
```

`ü` is not there.

So I guess that `set LC_ALL C` trick doesn't work well for the `read` builtin on Ubuntu.

**Comment by jakwings** [comment]
Also, I replace the whole function body of `__fish_urlencode` with `echo -n $HOME/T%C3%BCr; and return`, then cd to `$HOME/Tür` and <kbd>Ctrl</kbd><kbd>Shift</kbd><kbd>T</kbd> open a new tab, that works. So if `read` is replaced by a better function, everything will work normally.

**Comment by kballard** [comment]
That's really interesting. Given the following code:

```fish
echo -n T\u00FC | begin; set -l IFS ''; set -l LC_ALL C; read -a foo; end
string escape $foo
```

on OS X this prints:

```
T
Ã
¼
```

(Ã and ¼ are U+00C3 and U+00BC, i.e. the UTF-8 bytes interpreted as unicode codepoints)

On Ubuntu this prints:

```
T
```

It's just dropping the bytes.

**Comment by jakwings** [comment]
One more example to show who is dropping the bytes:

```fish
echo T\u00FCr > txt

begin
  set -lx LC_ALL C
  read -l chars < txt
  string escape $chars  #=> Tr
  string escape < txt  #=> T\Xc3\Xbcr
  string escape (/bin/cat txt)  #=> T\Xc3\Xbcr
  /bin/cat txt | string escape  #=> T\Xc3\Xbcr
end
```

PS: The issue title should be changed.

**Comment by jakwings** [comment]
Compared with the different result on OS X:

```
echo T\u00FCr > txt
begin
  set -lx LC_ALL C
  read -l chars < txt
  string escape $chars
  string escape < txt
  string escape (/bin/cat txt)
  /bin/cat txt | string escape
end | od -tx1z
```

output:

```
0000000 54 c3 bc 72 0a 54 c3 bc 72 0a 54 c3 bc 72 0a 54  >Tür.Tür.Tür.T<
0000020 c3 bc 72 0a                                      >ür.<
0000024
```

The behavior of `string escape` is strange. `LC_ALL=C` is not handled very well.

See another example:

```fish
begin; set -l LC_ALL C; echo T\u00FCr; end | xxd
```

on Ubuntu:
```
fish: Wide character 252 has no narrow representation
00000000: 5472 0a                                  Tr.
```

on OS X:
```
00000000: 54fc 720a                                T.r.
```

**Comment by krader1961** [comment]
I instrumented the code. The `mbrtowc()` call in `builtin_read()` is returning -1 with errno==EILSEQ for the chars with the high bit set (the UTF8 chars) which means an invalid multibyte sequence. Which, obviously, wouldn't seem to make any sense for the C locale. I tested with LC_ALL set to C, POSIX, and ISO-8859-1 and the results were identical.

Unfortunately, as this [StackOverflow thread](http://stackoverflow.com/questions/15649165/behavior-of-extended-bytes-characters-in-c-posix-locale) points out the standard allows both the OS X (BSD) and Linux (GNU) behaviors. I think we just need to check if `MB_CUR_MAX == 1` and bypass mbrtowc when that is true.

And I'm willing to bet the different handling of combining diacritical marks was a red-herring. We can revisit that difference if working around glibc's weird mbrtowc() behavior doesn't fix @TimWolla's problem.

**Comment by jakwings** [comment]
@krader1961 

> And I'm willing to bet the different handling of combining diacritical marks was a red-herring. 

That only happens on OS X (using HFS+ file system). And fish uses real file path. Bash is different, it keep what the user pass to `cd`, especially for symbolic links.

On Ubuntu you can create two directory "Tür" and "Tür": `mkdir T\u00FCr Tu\u0308r`
Fish can differentiate between them for `$PWD` and `pwd`.

**Comment by krader1961** [comment]
I have a fix for this specific issue that changes just two functions. However, there are multiple places that call mbrtowc() and its inverse wcrtomb(). There is also one location that calls the non-threadsafe wctomb(). It is going to require more thought to decide on the best overall solution to this problem. Or at least a solution that doesn't solve one very narrow problem while ignoring several related problems.

**Comment by krader1961** [comment]
I'm also adding this issue to the 2.3.0 milestone because it is essential for the already committed fix for setting the terminal title.

**Comment by krader1961** [comment]
Important question: How should wide characters be written in the C/POSIX locale?

While writing unit tests for the new logic I wrote a test like this

```
env LC_ALL=C ../test/root/bin/fish -c 'echo -n X\u01fdX' | xxd --plain
```

Knowing that my new code treats every character, wide or not, as if it were a single byte when serializing the string I knew it would output the equivalent of "X\xfdX" (with the \xfd being a literal byte of that value). Notice that the high byte (the \x01) of the wide char is omitted. I think that's the only sensible thing to do. However, I'm interested in what others think. Should it emit both literal bytes, MSB then LSB? Should it encode them as "\u01fd"? Before you say "yes" to the latter option consider the "u" with an umlaut that started this whole issue. It's wide representation is \xfc; a single byte that many (most?) people using the ASCII locale expect to be printed as a single byte having that value, not encoded as "\u00fc".

It is essentially impossible to do the right thing in all situations if we allow, as we do now, wide characters to be specified literally via the "\uXXXX" syntax when the locale is C. But disallowing that syntax in the C locale is likely to break things in ways far worse than simply dropping the high byte when printing those characters. The bottom line is that fish presumes it is operating in a UTF-8 locale. Anything we do to support the C/POSIX locales going to involve trade-offs.

**Comment by krader1961** [comment]
Please also watch this [video](https://www.youtube.com/watch?v=sgHbC6udIqc&feature=youtu.be) by Ned Batchelder.

**Comment by floam** [comment]
The presentation can also be read through [on his website](http://nedbatchelder.com/text/unipain.html).

**Comment by krader1961** [comment]
@jakwings pointed out something I thought I had already made clear: My proof of concept change does not make fish do the right thing when handed Unicode characters outside the code point range of 0 thru 255 in every possible scenario when the locale is C/POSIX. That is essentially impossible. My change is really meant to address just the very narrow problem identified by this issue that affects how the * __fish_urlencode* function behaves.

What exactly should fish do when presented with a "\uXXXX" character in a literal string when the locale is C/POSIX? Should it implicitly convert it to its UTF-8 representation before storing it internally? The answer is maybe yes. What my proof of concept change does is simply ignore everything but the LSB when printing that char. The best we can do is provide reasonable behavior that is easy to reason about.

**Comment by krader1961** [comment]
Should fish always use UTF-8 for input and output? That would solve some problems but create others. For example, the *__fish_urlencode* function would have to be rewritten to use a different, probably more expensive, means of encoding non-ASCII characters. And it is likely to surprise (i.e., cause problems) for someone using a ISO-8859-x locale. In other words the easy solution might be worse than the problem.

**Comment by oranja** [comment]
* Literal string codepoints in `LC_ALL=C`:
How about ignoring unicode escape sequences in C locale? Simply forget that they ever had any special meaning. The `\` escapes the `u` and the rest is just ASCII.

* UTF-8 input and output:
I don't think that forcing UTF-8 for input and output is the right way around. If fish really wants to be locale aware, it should always use the locale to figure out how to decode input and encode output. For example if the system uses a narrow Greek codepage, fish should be able to decode a `\0xD9` character (and a `\xd9` literal, maybe), as `Ω` (uppercase Omega). Literal unicode codepoints should not be treated specially in this case as well (no conversion to a unicode wide-char). Output should be encoded with a the same Greek codepage. Unicode codepoints, if can't be transposed to within the Greek codepage, should be replaced with a `?`, as @floam suggested.
Surely this is not nice, but I think that fish should assume that if a specific restrictive locale was chosen, the system will 

* Input and ouput with `LC_ALL=C`:
Read one byte at a time, and as you proposed, skip the call to `mbrtowc`, because of it's portability issues. I'd try to avoid any internal conversion to wide-chars. Only if there's no way around it, I would choose utf-8 for encoding back for output. I dare say that it's the first guess anyone would try.

As for the url-encode example, this is a special case where there is a standard [citation needed] that says that the string should be encoded to utf-8 and then escaped with literals. The end result is just ASCII, so it shouldn't be a problem with any locale.

Bottom line: Building a portable locale-aware system is tough. If an external dependency makes it any easier, and provides some magic string implementation where you can always know how is your string encoded, I'd take it without hesitation.

**Comment by krader1961** [comment]
I don't think ignoring the `\uXXXX` sequences is the right thing to do because that introduces the problem this is trying to solve. Namely, characters just disappearing. After reading the feedback on my change and thinking about it I've decided that the Python encode/decode "replace" on error behavior is preferable. Specifically, rather than dropping the character, or emitting just the low byte as my current change does, it's preferable to emit a question-mark.

**Comment by oranja** [comment]
Just for clarification: I did not mean ignore the entire `\uXXXX`, but rather read it as `\u`(=`u`) `X` `X` `X` `X`. Meaning not fooling the user that a C locale handles unicode code-points well. This is only in the context of string literals spelled directly in the terminal.
On the other hand it is possible that someone will copy a unicode character and paste it in the C locale terminal. In that case the treatment will be different than handling a `\uXXXX` literal with the same value (supposedly it will be read as whatever bytes make up its encoding in the terminal), and isn't this kind of inconsistency something to avoid? So perhaps indeed keep taking and converting unicode literals.

But... What I dislike about the surrogates strategy is that it results in data loss. In a `US-ASCII` locale, it may be acceptable, but as far as my understanding of the [uses of the C locale](http://unix.stackexchange.com/questions/87745/what-does-lc-all-c-do) goes, data loss seems wrong and incompatible with what people expect from this locale.
Under the C locale, files called `⁂` (\u2042) and `⅀` (\u2140) should be both survive a `ls | sort | uniq`, but with surrogates they will both be replaced with a `?` and only one `?` will remain of the two. To overcome this, fish can declare how Unicode is encoded internally (UCS4/UTF32/UTF8/etc) and then you can treat these filenames byte-wise and without data-loss. Personally I think this is the more accurate way.

Choosing an internal encoding also affects the previous point regarding unicode literals. Even if in a US-ASCII locale you choose to replace with surrogates, won't the number of `?` be affected by how many bytes are used to encode internally? So after reading, thinking and writing (which forces one to think even harder) about this matter, I conclude with two points:
1. IMO a first productive step would be to choose an internal encoding scheme, since I suspect that from that point on it will be easier to define a consistent behavior.
2. I believe that the `C` locale should accept every 8-bit character and be handled differently from a `US-ASCII` locale. The latter may replace out-of-range characters with `?`s, while the former must always read the string in its encoded form, as a single-byte array.

**Comment by jakwings** [comment]
@oranja 

> Just for clarification: I did not mean ignore the entire `\uXXXX`, but rather read it as `\u`(=u) `X` `X` `X` `X`. Meaning not fooling the user that a C locale handles unicode code-points well. This is only in the context of string literals spelled directly in the terminal.

Sounds like what the bash builtin `printf` is doing in the C locale.

```bash
export LC_ALL=C LANG=
printf '%b' '\u80'  #=> \u0080
printf '%b' '\U10000'  #=> \u00010000
export LC_ALL=en_US.UTF-8
printf '%b' '\U0001F41F'  #=> 🐟
```

> Under the C locale, files called ⁂ (\u2042) and ⅀ (\u2140) should be both survive a ls | sort | uniq, but with surrogates they will both be replaced with a ? and only one ? will remain of the two. To overcome this, fish can declare how Unicode is encoded internally (UCS4/UTF32/UTF8/etc) and then you can treat these filenames byte-wise and without data-loss.

`ls` is not a builtin function, and the encoding of filenames is independent of the shell. Only contents that are passed to the fish builtins get converted and filtered. It is the terminal (emulator)'s job to interpret the content of the screen buffer with a specified encoding.

**Comment by faho** [comment]
For some reason this wasn't closed via the commit.

**Comment by faho** [comment]
Sorry, that commit wasn't yet merged.

**Comment by TimWolla** [comment]
I can confirm that the issue is fixed for me, thanks!
